zkvm/src/signature/counterparty.rs

```diff
@@ -0,0 +1,109 @@
+use crate::errors::VMError;
+use crate::signature::multikey::Multikey;
+use crate::signature::VerificationKey;
+use crate::transcript::TranscriptProtocol;
+use curve25519_dalek::constants::RISTRETTO_BASEPOINT_POINT;
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use merlin::Transcript;
+use subtle::ConstantTimeEq;
+
+#[derive(Copy, Clone)]
+pub struct NoncePrecommitment([u8; 32]);
+
+#[derive(Copy, Clone, Debug)]
+pub struct NonceCommitment(RistrettoPoint);
+
+impl NonceCommitment {
+    pub fn new(commitment: RistrettoPoint) -> Self {
+        NonceCommitment(commitment)
+    }
+
+    pub fn precommit(&self) -> NoncePrecommitment {
+        let mut h = Transcript::new(b"MuSig.nonce-precommit");
+        h.commit_point(b"R", &self.0.compress());
+        let mut precommitment = [0u8; 32];
+        h.challenge_bytes(b"precommitment", &mut precommitment);
+        NoncePrecommitment(precommitment)
+    }
+
+    pub fn sum(commitments: &Vec<Self>) -> RistrettoPoint {
+        commitments.iter().map(|R_i| R_i.0).sum()
+    }
+}
+
+pub struct Counterparty {
+    pubkey: VerificationKey,
+}
+
+pub struct CounterpartyPrecommitted {
+    precommitment: NoncePrecommitment,
+    pubkey: VerificationKey,
+}
+
+pub struct CounterpartyCommitted {
+    commitment: NonceCommitment,
+    pubkey: VerificationKey,
+}
+
+impl Counterparty {
+    pub fn new(pubkey: VerificationKey) -> Self {
+        Counterparty { pubkey }
+    }
+
+    pub fn precommit_nonce(self, precommitment: NoncePrecommitment) -> CounterpartyPrecommitted {
+        CounterpartyPrecommitted {
+            precommitment,
+            pubkey: self.pubkey,
+        }
+    }
+}
+
+impl CounterpartyPrecommitted {
+    pub fn commit_nonce(
+        self,
+        commitment: &NonceCommitment,
+    ) -> Result<CounterpartyCommitted, VMError> {
+        // Check H(commitment) =? precommitment
+        let received_precommitment = commitment.precommit();
+        let equal = self.precommitment.0.ct_eq(&received_precommitment.0);
+        if equal.unwrap_u8() == 0 {
+            return Err(VMError::MuSigShareError {
+                pubkey: self.pubkey.0.to_bytes(),
+            });
+        }
+
+        Ok(CounterpartyCommitted {
+            commitment: *commitment,
+            pubkey: self.pubkey,
+        })
+    }
+}
+
+impl CounterpartyCommitted {
+    pub fn sign(
+        &self,
```

Comment:
Do we need to preserve the counterparty state after receiving the signature? If it gave us bad signature, we need to restart the whole protocol from scratch (since nonces are burned), and if it succeeded, we can return the share (and other data if needed), destroying the counterparty state.

Reply:
The only reason this takes `&self` instead of `self` is to help with the trouble I was having here:
https://github.com/interstellar/slingshot/pull/229#discussion_r267110317

(I would very much like to figure out a way to make this take `self` and still compile correctly!)

Reply:
done!